lib/active_merchant/billing/gateways/xpay.rb

```diff
@@ -30,77 +30,162 @@ def initialize(options = {})
+          gsub(%r(("cvv\\?":\\?")\d+), '\1[FILTERED]')
+      end
 
       private
 
-      def add_invoice(post, money, options) end
+      def add_invoice(post, amount, options)
+        currency = options[:currency] || currency(amount)
+        post[:order] = {
+          orderId: options[:order_id],
+          amount: localized_amount(amount, currency),
+          currency: currency
+        }.compact
+      end
+
+      def add_credit_card(post, payment_method)
+        post[:card] = {
+          pan: payment_method.number,
+          expiryDate: expdate(payment_method),
+          cvv: payment_method.verification_value
+        }
+      end
+
+      def add_payment_method(post, payment_method)
+        add_credit_card(post, payment_method) if payment_method.is_a?(CreditCard)
+      end
+
+      def add_customer_data(post, payment_method, options)
+        post[:order][:customerInfo] = {
+          cardHolderName: payment_method.name,
+          cardHolderEmail: options[:email]
+        }.compact
+      end
+
+      def add_address(post, options)
+        if options[:billing_address] || options[:address]
+          post[:order][:customerInfo][:billingAddress] = {
+            name: address[:name],
+            street: address[:address1],
+            additionalInfo: address[:address2],
+            city: address[:city],
+            postCode: address[:zip],
+            country: address[:country]
+          }.compact
+        end
+        if options[:shipping_address]
+          post[:order][:customerInfo][:shippingAddress] = {
+            name: address[:name],
+            street: address[:address1],
+            additionalInfo: address[:address2],
+            city: address[:city],
+            postCode: address[:zip],
+            country: address[:country]
+          }.compact
+        end
+      end
+
+      def add_recurrence(post, options)
+        post[:recurrence] = { action: options[:recurrence] || 'NO_RECURRING' }
+      end
+
+      def add_exemptions(post, options)
+        post[:exemptions] = options[:exemptions] || 'NO_PREFERENCE'
+      end
+
+      def add_3ds_params(post, options)
+        post[:threeDSAuthData] = { threeDSAuthResponse: options[:three_ds_auth_response] }.compact
+        post[:operationId] = options[:operation_id] if options[:operation_id]
+      end
 
-      def add_payment_method(post, payment_method) end
+      def add_auth_purchase_params(post, amount, payment_method, options)
+        add_invoice(post, amount, options)
+        add_payment_method(post, payment_method)
+        add_customer_data(post, payment_method, options)
+        add_address(post, options)
+        add_recurrence(post, options) unless options[:operation_id]
+        add_exemptions(post, options)
+        add_3ds_params(post, options)
+      end
 
       def add_reference(post, authorization) end
 
       def add_auth_purchase(post, money, payment, options) end
 
+      def parse(body)
```

Comment:
💬 Style/Comment:

Maybe a Dry approach for the same result;

```ruby
      def parse(body = '{}')
        JSON.parse(body)
      end

```

Reply:
Noted! Thanks!